### DISCUSSION TITLE
How to set a cell's visibility based on graph state?

### DISCUSSION THREAD

**Discussion by sdgoldstein** [maxGraph/maxGraph#179] [Q&A] [answered]
I've been working on moving a project from mxGraph to maxGraph.  

The issue I've run into is that some of the cell visibility methods on the Graph and GraphModel types have been removed.  The use case that I'm not sure how to implement now is this:

"Show or hide a cell based on a given application state"

One way I can see to implement this now, is when the state changes, go through every cell in the graph and set the visible flag accordingly.  But, this is wasteful.

In mxGraph, one could override Graph.isCellVisible(cell) or GraphModel.isVisible(cell) which would be called during layout of the graph through the GraphLayout.isVertexIgnored(cell) function.

Let me know if I'm simply not seeing a way to achieve this in maxGraph.

Thanks!

**Comment by tbouffard** [comment]
Hi @sdgoldstein and sorry for the late answer.

`mxGraph.isCellVisible(cell)` or `mxGraphModel.isVisible(cell)` methods were probably removed during the code migration. I didn't find any reference to this in various discussions in the past, like in #8, #18 or #51. I guess that a deep search in the git history may give us more insights.  
Maybe @mcyph or @junsikshim have more information on this topic.

During migration, it was hard to figure out what was real extension points or extra design. For the scenario you mention, the methods may have been removed by mistake.

You also mentioned "Show or hide a cell based on a given application state". Is the state stored somewhere in the maxGraph/mxGraph view state or outside maxGraph? If it is stored outside, you may show/hide a cell using `GraphDataModel.setVisible` (it previously existed in mxGraph). Couldn't this work for you? 

https://github.com/maxGraph/maxGraph/blob/d33769fad5e6c2ee1dd216939e07deb4745381b7/packages/core/src/view/GraphDataModel.ts#L910-L922


Otherwise, having a way to override `Graph.isCellVisible(cell)` should be enough as you need to access to the view state.
Prior creating an issue for this, I suggest you clarify your use case with us.

Thanks

**Reply by mcyph** [reply]
I can confirm as part of the migration many methods were migrated into the `Cell` type itself (in this case moved into `Cell.isVisible()`) because I thought it generally would be easier working with objects than as state variables. 

I can't remember the exact details of this, and I wouldn't expect having this behavior in the graph model would be significantly more efficient (maxGraph probably goes through every cell and calls methods like these for each every time it renders anyway, for instance). 

But I think this issue raises a good question about how to achieve the same kind of behavior graph-wide. It may be that adding this functionality back into GraphModel, or allowing overriding the Cell class to add custom behavior may be an option. I'll need to look thru the code to refresh my memory.

**Comment by sdgoldstein** [comment]
Thank you both for the replies!  Apologies that my use case wasn't clear.  In my scenario, the state is stored in the Value object in the cell itself.  Whether or not a cell is visible or not is based on `Value.getType()` where in a particular app view, cells of a certain type should be hidden.

@mcyph I think moving the `isVisible()` method to the cell itself makes sense.  However, I don't know how to extend the Cell class to create a custom `isVisible()` implementation.  The only way I know how to add a cell is through `Graph.insertVertex()`, which returns the Cell.  If I could override `Cell.isVisible()` to invoke `Value.getType()`, that would work well.

Where I used this previously is in a custom GraphLayout.  During the layout as the tree is being traversed, I call Cell.isVisible() to decide whether or not to render the cell.

@tbouffard - To answer your question, I could use `GraphDataModel.setVisible()`.  However, I believe that would require two traversals of the tree.  During the first, when my user selects a particular app view, I would have to traverse the tree to call `GraphDataModel.setVisible()` for each cell.  Then, I would have to rerender the tree, which would traverse the tree again during layout.

**Reply by mcyph** [reply, marked as answer]
Yeah absolutely - I think there should be a means of globally overriding `Cell` methods. 

As can be seen at this discussion: https://github.com/maxGraph/maxGraph/discussions/151 we have plans to allow overriding other classes in the `plugins` argument, and I think that `Cell` should also be added to this.
